packages/app-desktop/gui/MainScreen/commands/perNotebookSortOrder.ts

```diff
@@ -0,0 +1,212 @@
+import CommandService, { CommandContext, CommandDeclaration, CommandRuntime } from '@joplin/lib/services/CommandService';
+import Setting from '@joplin/lib/models/Setting';
+import { State } from '@joplin/lib/reducer';
+import app from '../../../app';
+import eventManager from '@joplin/lib/eventManager';
+import {
+	notesSortOrderFieldArray, NOTES_SORT_ORDER_SWITCH, SETTING_FIELD, SETTING_REVERSE, SETTING_PER_FIELD_REVERSAL_ENABLED,
+} from './notesSortOrderSwitch';
+import { _ } from '@joplin/lib/locale';
+
+
+export const PER_NOTEBOOK_SORT_ORDER = 'perNotebookSortOrder';
+export const SETTING_PER_NOTEBOOK_SORT_ORDER_ENABLED = 'notes.perNotebookSortOrderEnabled';
+
+let previousNotebookId: string = null;
+let ownSortOrders: { [key: string]: any } = null;
+const sharedSortOrder: { [key: string]: any } = {
+	field: 'user_updated_time',
+	reverse: true,
+	user_updated_time: true,
+	user_created_time: true,
+	title: false,
+	order: false,
+};
+
+export function selectedNotebookHasOwnSortOrder() {
+	return hasOwnSortOrder(getSelectedNotebookId());
+}
+
+export function hasOwnSortOrder(notebookId: string) {
+	return notebookId && ownSortOrders && ownSortOrders.hasOwnProperty(notebookId + SUFFIX_FIELD);
+}
+
+export const declaration: CommandDeclaration = {
+	name: PER_NOTEBOOK_SORT_ORDER,
+	label: () => _('Toggle own sort order'),
+};
+
+export const runtime = (): CommandRuntime => {
+	loadOwnSortOrders();
+	loadSharedSortOrder();
+	eventManager.appStateOn('notesParentType', onNotebookSelectionMayChange);
+	eventManager.appStateOn('selectedFolderId', onNotebookSelectionMayChange);
+
+	return {
+		enabledCondition: 'oneFolderSelected',
+
+		execute: async (context: CommandContext, notebookId?: string, own?: boolean) => {
+			let targetId = notebookId;
+			const selectedId = getSelectedNotebookId(context.state);
+			if (!targetId) {
+				targetId = selectedId; // default: selected notebook
+				if (!targetId) return;
+			}
+			const targetOwn = hasOwnSortOrder(targetId);
+			let newOwn;
+			if (typeof own == 'undefined') {
+				newOwn = !targetOwn; // default: toggling
+			} else {
+				newOwn = !!own;
+				if (newOwn === targetOwn) return;
+			}
+			if (newOwn) {
+				let field: string, reverse: boolean;
+				if (!hasOwnSortOrder(selectedId)) {
+					field = getCurrentField();
+					reverse = getCurrentReverse();
+				} else {
+					field = sharedSortOrder.field;
+					if (perFieldReversalEnabled()) {
+						reverse = sharedSortOrder[field];
+					} else {
+						reverse = sharedSortOrder.reverse;
+					}
+				}
+				setOwnSortOrder(targetId, field, reverse);
+			} else {
+				deleteOwnSortOrder(targetId);
+			}
+		},
+	};
+};
+
+function onNotebookSelectionMayChange() {
```

Comment:
This should not be here - there has to be some existing code that does something when folder selection change, so please move it over there. That way, you also won't need to include the app package in the comment (creates a dependency loop).

Reply:
First, the dependency on 'app' could be removed from this command without any redesign. 5d0b0e3

Before we proceed further, I want to clarify your intentions.

> This should not be here - there has to be some existing code that does something when folder selection change, so please move it over there. 

Which is your suggestion intend, only removing dependency on `app` or that such a functionality should be moved toward the core?

If the former, everything is done.

If the latter, I have some questions. The first is where is it moved to? Around lib/reducer.ts, or around app-desktop/app.reducer.ts, or in some React component (maybe, NoteListXXX)? I'm not sure the features play a role in the core rather than app-desktop.

Another question is how to capture changes of states/settings. First, modifying lib/reducer.ts is definite and effective, but it will worse a certain degree of the maintainability, because changing codes of folerSelectionId and notesParentType are scattered. In another side, I don't have a good solution to modify app-desktop/app.reducer.ts. Finally, componentDidUpdated(), useEffect() and appStateOn() are handy to implement the feature.